Allow filtering in value relation widget when allowing multiple selections

uclaros opened the pull request:
## Description
When using the Value Relation widget on a form and having _Allow multiple selections_ enabled, the items were added to a `QTableWidget` with the specified number of columns.
This PR uses a new widget instead that also offers a `QgsFilterLineEdit` allowing filtering the displayed table's items.
The line edit widget is only displayed when both  _Allow multiple selections_ and _Use completer_ are enabled.
When _Use completer_ is disabled, the functionality should be unchanged.

Here is a sample with two fields, one with 5 columns and completer, one with 3 columns and no completer, using lots of bird names as values:



https://user-images.githubusercontent.com/11358178/207796644-c50d713c-2786-43a0-8949-0fa388281783.mp4



Fixes #50371
Fixes unreported bug: after stopping editing on the attribute table while still editing a cell, the next cell appears editable, related to #8524
<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

uclaros committed:
custom filtered table implementation

uclaros committed:
remove hack when stopping editing

uclaros committed:
add tests, fix existing test

troopa81 commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h:
```diff
@@ -16,17 +16,98 @@
+    /**
+     * Set all table items to \a enabled.
+     */
+    void setEnabledTable( const bool enabled );
```
nitpick: why not just setEnabled ?

uclaros replied:
`QWidget` has a `setEnabled` non virtual slot and I didn't want to shadow it / cause any confusion

troopa81 replied:
Right!

troopa81 reviewed (commented):
Nice feature :+1: 

Have you considered using QAbstractItemModel instead of a QTableWidget since all the model logic is within mCache ? It will avoid allocating QTableWidget every time filter string changes, no?

Besides that, is it possible to make the space between lines constant so it doesn't grow when we filter out items ?

uclaros commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h:
```diff
@@ -16,17 +16,98 @@
+ * This is a helper widget for QgsValueRelationWidgetWrapper
+ * This widget is a QTableWidget showing checkable items, with an optional QgsFilterLineEdit on top that allows filtering the table's items
+ */
+class QgsFilteredTableWidget : public QWidget
```
The thing is that some of its members are used in the tests: `mTableWidget->item()`, `mTableWidget->rowCount()` and `mTableWidget->mSearchWidget->setText()`.
I would need to create some (private) methods in `QgsValueRelationWidgetWrapper` to handle those, which is a bit ugly. What do you think?

troopa81 replied:
I agree, I didn't think about the tests. No need to change anything

uclaros commented:
> Have you considered using QAbstractItemModel instead of a QTableWidget since all the model logic is within mCache ? It will avoid allocating QTableWidget every time filter string changes, no?

I though about it a bit at first, but I found confusing the part that the model (which can be filtered) should be one dimensional and should be displayed in a two dimensional QTableView. It would probably need some extra intermediate proxy model to take care of the distribution of items to columns... so I went with reusing the simpler existing QTableWidget/QTableWidgetItem
I also though of adding a singleshot timer delay before applying the filter string to limit the table rebuilds between keystrokes, but did not see a need for it while testing with thousands of values.

> Besides that, is it possible to make the space between lines constant so it doesn't grow when we filter out items ?

Styling (and the rest of the functionality) was left unchanged, so this is how the table also looks when the total number of items is small.

A potential future improvement would be a context menu to offer check/uncheck all/filtered

uclaros committed:
@cond in implementation

uclaros committed:
replace deprecated setMargin() with setContentsMargins()

troopa81 commented:
> I though about it a bit at first, but I found confusing the part that the model (which can be filtered) should be one dimensional and should be displayed in a two dimensional QTableView. It would probably need some extra intermediate proxy model to take care of the distribution of items to columns

I was more thinking about a 2-dimensionnal model taking care about the filtering, (with *data()* method containing quite the same logic of the *filterStringChanged* method), not a 1-dimensionnal model with proxysortfiltermodel which, I agree would be too complicated.

> Styling (and the rest of the functionality) was left unchanged, so this is how the table also looks when the total number of items is small.

OK

uclaros commented:
> I was more thinking about a 2-dimensionnal model taking care about the filtering, (with _data()_ method containing quite the same logic of the _filterStringChanged_ method),

hmm, sorry, I'm still not sure how filtering would work without a filter proxy model that would also translate filtered indexes to original indexes.

uclaros committed:
remove redundant python test:
This test needed to access private QgsValueRelationWidgetWrapper members
which is possible and done on the existing cpp test.

uclaros committed:
make clang-tidy happy

troopa81 commented:
> hmm, sorry, I'm still not sure how filtering would work without a filter proxy model that would also translate filtered indexes to original indexes.

- one QAbstractItemModel
- data() returning elements according to the same logic that the one describe in filterStringChanged method,
- setData to modify mCache on check/uncheck
- rowCount/colCount where computation is based on the config parameters
- send signal layoutChanged() when the content of the filter line edit has changed 

And I think that's pretty it! 

QTableWidget is for display facilities and the Qt documentation state that 

> If you want a table that uses your own data model you should use [QTableView](https://doc.qt.io/qt-6/qtableview.html) rather than this class.

And I think that's kind of the situation here.

uclaros commented:
> * rowCount/colCount where computation is based on the config parameters

rowCount is variable and you do not know it until all of mCache is checked against the filter string

> * data() returning elements according to the same logic that the one describe in filterStringChanged method,

filtering logic cannot live within the data() method. The accepted index range depends on the filtered results, hence the need for a proxy model.

troopa81 commented:
@uclaros Yes, you're right. Thanks for clarification!

I'm OK to merge, I'm restarting the CI because I don't get why the build artifact is missing
